sale.py

```diff
@@ -57,3 +50,31 @@ def get_shipment_state(self):
                 if any(s for s in self.shipments if s.state in ('done', 'cancelled')):
                     return 'sent'
         return super(Sale, self).get_shipment_state()
+
+    @classmethod
+    def process(cls, sales):
+        pool = Pool()
+        SaleLine = pool.get('sale.line')
+        ShipmentOut = pool.get('stock.shipment.out')
+
+        super().process(sales)
+
+        for sale in sales:
+            if (sale.shipment_state == 'sent'
+                    and sale.remaining_stock == 'manual'):
+                # cancel shipments and add to moves ignored
+                shipments = [s for s in sale.shipments
+                    if s.state not in ('cancelled', 'done')]
+                ShipmentOut.cancel(shipments)
+                for line in sale.lines:
+                    moves = []
+                    skips = set(line.moves_ignored)
+                    skips.update(line.moves_recreated)
+                    for move in line.moves:
+                        if move.state == 'cancelled' and move not in skips:
+                            moves.append(move.id)
+                    if not moves:
+                        continue
+                    SaleLine.write([line], {
+                            'moves_ignored': [('add', moves)],
+                            })
```

Comment:
Why not using instances .save() instead of write. Also better do all of them in batch at the very end of the process method so a single `save()` is enough for all lines in all sales.

Reply:
Same as [HandleShipmentException](https://github.com/tryton/tryton/blob/3fc39d0433e68354fe6852402a8723364fab759e/modules/sale/sale.py#L2330)